### PR TITLE
Homework 3

### PR DESCRIPTION
**Pull request by tribaeka**

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+version: '3.7'
```
Comment (kuzanatoliy) [inline docker-compose.yml:1]: You can just add script something like it
```"docker run --rm -it -p 5432:5432 --name gamesbe -e POSTGRES_PASSWORD=root -e POSTGRES_USER=root -e POSTGRES_DB=games.db postgres:13"```

File: homework-2/middewares/userMiddlewares.ts [hunk 2]
```diff
@@ -27,26 +27,40 @@ export function getUserByIdHandler(req: Request<IIdParam>, res: Response): void
...
+    const autoSuggests = await UserService.getAutoSuggestUsers(req.query.loginSubstring, req.query.limit);
     res.send(autoSuggests);
 }
 
-export function addUserHandler(req: Request<Record<string, unknown>, unknown, IUser>, res: Response): void {
+export async function addUserHandler(
+    req: Request<Record<string, unknown>, unknown, IUserAttrs>,
+    res: Response
+): Promise<void> {
     const errors = validationResult(req);
```
Comment (kuzanatoliy) [inline homework-2/middewares/userMiddlewares.ts:43]: I would use separate handler for validation

File: homework-2/models/index.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import { Sequelize } from "sequelize";
+import { userFactory } from "./userModel";
+import { IDB } from "../interfaces/IDB";
+
+const sequelize  = new Sequelize(
+    'postgres',
```
Comment (kuzanatoliy) [inline homework-2/models/index.ts]: Bad practice to save credentials into code. It is better to use environment.
https://www.npmjs.com/package/dotenv

File: homework-2/services/userService.ts [hunk 4]
```diff
@@ -1,83 +1,57 @@
...
-        return this.usersList;
+class UserService {
+    getAllUsers(): Promise<IUserAttrs[]> {
+        return db.User.findAll()
     }
 
-    getUserById(id: string): IUser | null {
-        return this.usersList.find(user => user.id === id) || null;
+    getUserById(id: string): Promise<IUserModel | null> {
+        return db.User.findByPk(id);
```
Comment (kuzanatoliy) [inline homework-2/services/userService.ts]: Please, use toJson method. Service hasn't return dbModel

File: homework-2/services/userService.ts [hunk 5]
```diff
@@ -1,83 +1,57 @@
...
     }
 
-    addUser(user: IUser): string {
-        user.id = user.id || uuidv4();
-        this.usersList.push(user);
+    async addUser(user: IUserAttrs): Promise<number> {
```
Comment (kuzanatoliy) [inline homework-2/services/userService.ts:24]: I would add method getUserByLogin and move checking part into controller level

File: homework-2/services/userService.ts [hunk 6]
```diff
@@ -1,83 +1,57 @@
...
+        const users = await this.getAllUsers();
+
+        return users
+            .sort(UserService.compareByLogin)
             .filter(user => user.login.includes(loginSubstring))
             .slice(0, limit);
     }
 
-    private compareByLogin(u1: IUser, u2: IUser): number {
+    private static compareByLogin(u1: IUserAttrs, u2: IUserAttrs): number {
```
Comment (kuzanatoliy) [inline homework-2/services/userService.ts]: Please use this string method
https://www.w3schools.com/jsref/jsref_localecompare.asp

File: docker-compose.yml (hunk 1 above)
Comment (tribaeka) [reply docker-compose.yml:1]: You're right but as I'm also using 'volumes', the run command will look too long. That's why I created a docker-compose file.

File: homework-2/middewares/userMiddlewares.ts (hunk 2 above)
Comment (tribaeka) [reply homework-2/middewares/userMiddlewares.ts:43]: todo for q&a session

File: homework-2/models/index.ts (hunk 3 above)
Comment (tribaeka) [reply homework-2/models/index.ts]: Fixed, thanks

File: homework-2/services/userService.ts (hunk 4 above)
Comment (tribaeka) [reply homework-2/services/userService.ts]: Fixed, thanks

File: homework-2/services/userService.ts (hunk 5 above)
Comment (tribaeka) [reply homework-2/services/userService.ts:24]: Done

File: homework-2/services/userService.ts (hunk 6 above)
Comment (tribaeka) [reply homework-2/services/userService.ts]: Done, thanks

**Pull request merged**